### PR TITLE
Action cancel/abort handling on uniter

### PR DESCRIPTION
**Pull request by hpidcock**
### Checklist

 - [ ] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [ ] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [ ] Do comments answer the question of why design decisions were made?

----

## Action cancel/abort handling on uniter

- Extend k8s exec framework to support signals and cancelation (with
kill).
- Uniter reacts to action change to check for aborting status

## QA steps

Create a charm with an action that sleeps for a few minutes.
Deploy charm.
`juju cancel-action <action-id>`
Action should go to aborting status.
Then should go to aborted status.

## Documentation changes

N/A

## Bug reference

N/A

### CODE REVIEW

File: worker/uniter/runner/runner.go [hunk 1]
```diff
@@ -488,9 +491,12 @@ func (runner *runner) runCharmHookOnLocal(hookName string, env []string, charmLo
 	// separately to pass back.
 	var actionErr io.Reader
 	var hookErrLogger *charmrunner.HookLogger
-	_, err = runner.context.ActionData()
-	runningAction := err == nil
+	var cancel <-chan struct{}
```
Comment (wallyworld) [inline worker/uniter/runner/runner.go]: runCharmHookOnLocal() seems a bit misnamed.
Perhaps rename as a driveby.
eg runCharmProcessOnLocal
Also runCharmHookOnRemote

File: caas/kubernetes/provider/exec/exec.go [hunk 2]
```diff
@@ -167,11 +180,88 @@ func (c client) exec(opts ExecParams, cancel <-chan struct{}) error {
...
+			}
+			// Trigger SIGKILL
+			timer = time.After(gracefulKillDelay)
+		case <-kill:
+			err := sendSignal(syscall.SIGKILL, true)
+			if err != nil {
+				return errors.Annotatef(err, "send signal %d failed", syscall.SIGKILL)
+			}
+			// Retry SIGKILL.
+			timer = time.After(sigkillRetryDelay)
```
Comment (wallyworld) [inline caas/kubernetes/provider/exec/exec.go]: I think we'd want an escape hatch out of the loop, eg after X tries or X time.

**Comment by wallyworld** [comment]
[Action abortion](https://trello.com/c/hJ0cM7NN/1039-action-abortion)

File: caas/kubernetes/provider/exec/exec.go [hunk 3]
```diff
@@ -23,6 +27,11 @@ import (
 
 var logger = loggo.GetLogger("juju.kubernetes.provider.exec")
 
+const sigkillRetryDelay = 100 * time.Millisecond
+const gracefulKillDelay = 10 * time.Second
+
+var randomString = utils.RandomString
```
[the remark this answers is not in the document]
Comment (hpidcock) [reply caas/kubernetes/provider/exec/exec.go]: randomPrefixFunc isn't exported and we don't use the broker in the exec tests. Don't think it makes sense to rejig this.

File: worker/uniter/op_callbacks.go [hunk 4]
```diff
@@ -108,6 +108,14 @@ func (opc *operationCallbacks) FailAction(actionId, message string) error {
 	return err
 }
 
+func (opc *operationCallbacks) ActionStatus(actionId string) (string, error) {
+	if !names.IsValidAction(actionId) {
+		return "", errors.Errorf("invalid action id %q", actionId)
```
[the remark this answers is not in the document]
Comment (hpidcock) [reply worker/uniter/op_callbacks.go]: NotValidf

**Comment by hpidcock** [comment]
$$merge$$
